### PR TITLE
Material palette

### PR DESCRIPTION
**Pull request by geralddejong**

### CODE REVIEW

File: src/experiment.rs [hunk 1]
```diff
@@ -95,7 +95,7 @@ impl Experiment {
                 let new_pulls = self.fabric.measures_to_pulls(*strain_threshold);
                 self.fabric = self.frozen_fabric.take().unwrap();
                 for (alpha_index, omega_index, ideal) in new_pulls {
-                    self.fabric.create_interval(alpha_index, omega_index, Link::Pull { ideal });
+                    self.fabric.create_interval(alpha_index, omega_index, Link::Pull { ideal, material: 1 });
```
Comment (mitchelljustin) [inline src/experiment.rs:98]: Extract material "1" to a meaningful const

File: src/fabric/mod.rs [hunk 2]
```diff
@@ -46,14 +46,22 @@ impl Default for Fabric {
...
+            pull_material: vec![
+                Material {
+                    stiffness: 1.0,
+                    mass: 0.1,
+                },
+                Material {
+                    stiffness: 0.7,
+                    mass: 0.1,
+                },
+            ],
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:64]: Why not enum? Are these going to be added to at runtime in the future?

**Review by mitchelljustin — CHANGES_REQUESTED** [review]
Justify your use of vecs and indices or change to an enum/consts

File: src/fabric/interval.rs [hunk 3]
```diff
@@ -41,8 +41,7 @@ pub struct Material {
 pub struct Interval {
     pub alpha_index: usize,
     pub omega_index: usize,
-    pub role: Role,
-    pub material: Material,
+    pub material: usize,
```
Comment (mitchelljustin) [inline src/fabric/interval.rs:44]:
```suggestion
    pub material_id: usize,
```
But maybe that's just me

File: src/fabric/interval.rs [hunk 4]
```diff
@@ -115,19 +112,17 @@ impl Interval {
             }
         };
         let real_length = self.length(joints);
-        self.strain = match self.role {
+        let Material{ role, stiffness, mass } = materials[self.material];
```
Comment (mitchelljustin) [inline src/fabric/interval.rs:115]:
```suggestion
        let Material { role, stiffness, mass } = materials[self.material];
```

File: src/fabric/mod.rs [hunk 5]
```diff
@@ -46,14 +45,18 @@ impl Default for Fabric {
...
+                    role: Push,
+                    stiffness: 3.0,
+                    mass: 1.0,
+                },
+                Material {
+                    role: Pull,
+                    stiffness: 1.0,
+                    mass: 0.1,
+                },
+            ],
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:59]: Better to extract this to a const array and call .into() on it here.

```rust
const DEFAULT_MATERIALS: [Material; 2] = [
    Material {
        role: Push,
        stiffness: 3.0,
        mass: 1.0,
    },
    Material {
        role: Pull,
        stiffness: 1.0,
        mass: 0.1,
    },
];
```

Then

```suggestion
            materials: DEFAULT_MATERIALS.into(),
```

File: src/fabric/mod.rs [hunk 6]
```diff
@@ -246,9 +231,17 @@ pub struct UniqueId {
...
+    material: usize,
+}
+
+impl Link {
+    pub fn push(ideal: f32) -> Self {
+        Self { ideal, material: 0 }
+    }
+
+    pub fn pull(ideal: f32) -> Self {
+        Self { ideal, material: 1 }
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:245]: Extract these indices to consts

File: src/fabric/mod.rs [hunk 7]
```diff
@@ -246,9 +231,17 @@ pub struct UniqueId {
...
 
 #[derive(Clone, Debug, Copy)]
-pub enum Link {
-    Push { ideal: f32 },
-    Pull { ideal: f32 },
-    PullStiffness { ideal: f32, stiffness_factor: f32 },
-    Measure { length: f32 },
+pub struct Link {
+    ideal: f32,
+    material: usize,
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:236]:
```suggestion
    material_id: usize,
```
Same suggestion

File: src/fabric/vulcanize.rs [hunk 8]
```diff
@@ -1,24 +1,39 @@
...
+use crate::fabric::interval::{Interval, Material, Role};
+use crate::fabric::Link;
 
 const ROOT3: f32 = 1.732_050_8;
-const BOW_TIE_STIFFNESS_FACTOR: f32 = 0.7;
 const BOW_TIE_PUSH_LENGTH_FACTOR_HEXAGON: f32 = 0.2;
 const BOW_TIE_PUSH_LENGTH_FACTOR_OCTAGON: f32 = 0.12;
 
 impl Fabric {
+    const BOW_TIE_MATERIAL: usize = 1;
```
Comment (mitchelljustin) [inline src/fabric/vulcanize.rs:12]: I think this is wrong? Isn't this supposed to be `2`?

File: src/fabric/mod.rs [hunk 9]
```diff
@@ -186,9 +170,10 @@ impl Fabric {
     pub fn prepare_for_pretensing(&mut self, push_extension: f32) {
         for interval in self.intervals.values_mut() {
             let length = interval.length(&self.joints);
-            interval.span = match interval.role {
+            let Material { role, .. } = self.materials[interval.material];
+            interval.span = match role {
```
Comment (mitchelljustin) [inline src/fabric/mod.rs:174]:
```suggestion
            interval.span = match self.materials[interval.material].role {
```
Overkill IMO

File: src/scene.rs [hunk 10]
```diff
@@ -225,13 +225,11 @@ struct FabricVertex {
 impl FabricVertex {
     pub fn for_interval(interval: &Interval, fabric: &Fabric, strain_threshold: f32) -> [FabricVertex; 2] {
         let (alpha, omega) = interval.locations(&fabric.joints);
-        let color = match interval.role {
+        let Material { role, .. } = fabric.materials[interval.material];
+        let below_threshold = interval.strain < strain_threshold;
+        let color = match role {
```
Comment (mitchelljustin) [inline src/scene.rs:230]:
```suggestion
        let color = match fabric.materials[interval.material].role {
```
IMO

File: src/experiment.rs [hunk 11]
```diff
@@ -93,14 +93,15 @@ impl Experiment {
...
-                self.start_pretensing()
+            AddPulls { .. } => {
+                unimplemented!();
+                // self.add_pulls = None;
+                // let new_pulls = self.fabric.measures_to_pulls(*strain_threshold);
+                // self.fabric = self.frozen_fabric.take().unwrap();
+                // for (alpha_index, omega_index, ideal) in new_pulls {
+                //     self.fabric.create_interval(alpha_index, omega_index, Link::Pull { ideal, material: 1 });
+                // }
+                // self.start_pretensing()
```
Comment (mitchelljustin) [inline src/experiment.rs:104]: What's the story here?

**Review by mitchelljustin — CHANGES_REQUESTED** [review]
A few more suggestions

File: src/fabric/mod.rs (hunk 7 above)
Comment (geralddejong) [reply src/fabric/mod.rs:225]: I'd consider it an index and not an id (identifier) and I don't want to use material_index everywhere.

File: src/experiment.rs (hunk 11 above)
Comment (geralddejong) [reply src/experiment.rs:104]: Still must figure out how to do this now.  Thought I'd keep the comment for reference.

File: src/fabric/vulcanize.rs [hunk 12]
```diff
@@ -1,24 +1,39 @@
...
 impl Fabric {
+    const BOW_TIE_MATERIAL: usize = 2;
+
+    pub fn default_bow_tie() -> Self {
+        let mut fabric = Fabric::default();
+        fabric.materials.push(Material {
+            role: Role::Pull,
+            stiffness: 0.7,
+            mass: 0.1,
+        });
```
Comment (mitchelljustin) [inline src/fabric/vulcanize.rs:20]: Extract this to `const BOW_TIE_MATERIAL`

File: src/fabric/vulcanize.rs [hunk 13]
```diff
@@ -35,32 +50,20 @@ impl Fabric {
...
         }
     }
 
-    pub fn max_measure_strain(&self) -> f32 {
-        self.interval_measures()
-            .map(|Interval { strain, .. }| strain)
+    pub fn max_strain(&self, target_material: usize) -> f32 { // todo: make this min/max
+        self.interval_values()
+            .filter_map(|&Interval { strain, material, .. }|
+                (material == target_material).then_some(strain))
```
Comment (mitchelljustin) [inline src/fabric/vulcanize.rs:62]: Love to see it

**Review by mitchelljustin — APPROVED** [review]
Cool!

**Pull request merged**